solutions/cses-1625.mdx

```diff
@@ -0,0 +1,317 @@
+slow (Java takes around 1.2 seconds for the worst case). Thus, we will need to
+further optimize our search.
+
+### Optimization 4
+
+If the path creates a dead end that is not the bottom left corner, either the
+path will fail to visit all squares (the path may stop at the dead end or pass
+over it, sealing a square off) or the path will end in the wrong location. Thus
+we want to avoid creating dead ends. For example, if the square to the left of
+our current location is blocked on three sides (including our current location),
+then the next step must be to the left in order to avoid creating a dead end.
+After this optimization, the program runs in under 1 second.
+
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+using namespace std;
+
+const int dirLen = 4;
+int dr[dirLen] = {-1, 0, 1, 0};
+int dc[dirLen] = {0, 1, 0, -1};
+const int pathLen = 48;
+int p[pathLen];
+const int gridSize = 9;
+bool onPath[gridSize][gridSize];
+
+int tryPath(int pathIdx, int curR, int curC)
+{
+	if ((onPath[curR][curC - 1] && onPath[curR][curC + 1])
+			&& (!onPath[curR - 1][curC] && !onPath[curR + 1][curC])) return 0;
+	if ((onPath[curR - 1][curC] && onPath[curR + 1][curC])
+					&& (!onPath[curR][curC - 1] && !onPath[curR][curC + 1])) return 0;
+
+	if (curR == 7 && curC == 1) { // reached endpoint before visiting all
+		if (pathIdx == 48) {
+			return 1;
+		}
+		return 0;
+	}
+
+	if (pathIdx == 48) {
+		return 0;
+	}
+
+	int ret = 0;
+	onPath[curR][curC] = true;
+
+	// turn already determined:
+	if (p[pathIdx] < 4) {
+		int nxtR = curR + dr[p[pathIdx]];
+		int nxtC = curC + dc[p[pathIdx]];
+		if (!onPath[nxtR][nxtC]) {
+			ret += tryPath(pathIdx + 1, nxtR, nxtC);
+		}
+	} /* // this next bit is not necessary for C++, but I left it in for reference
+	else if ((curC > 2) && onPath[curR][curC - 2]
+			&& (onPath[curR - 1][curC - 1] || onPath[curR + 1][curC - 1])
+			&& (!onPath[curR][curC - 1]))
+	{ // potential dead end on the left:
+		int nxtR = curR;
+		int nxtC = curC - 1;
+		ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	}
+	else if ((curC < 6) && onPath[curR][curC + 2]
+			&& (onPath[curR - 1][curC + 1] || onPath[curR + 1][curC + 1] )
+			&& (!onPath[curR][curC + 1]))
+	{ // potential dead end on the right:
+		int nxtR = curR;
+		int nxtC = curC + 1;
+		ret += tryPath(pathId x +1, nxtR, nxtC);
+	}
+	else if ((curR > 2) && onPath[curR - 2][curC] && onPath[curR - 1][curC - 1]
+			&& (!onPath[curR - 1][curC]))
+	{ // potential dead end upwards
+		int nxtR = curR - 1;
+		int nxtC = curC;
+		ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	} */
+	else { // iterate through all four possible turns
+		for (int i = 0; i < 4; i++) {
+			int nxtR = curR + dr[i];
+			int nxtC = curC + dc[i];
+			if (onPath[nxtR][nxtC]) continue;
+			ret += tryPath(pathIdx + 1, nxtR, nxtC);
+		}
+	}
+	// reset and return
+	onPath[curR][curC] = false;
+	return ret;
+}
+
+int main()
+{
```

Comment:
again, be consistent- either put all braces on a newline or all braces on the same line